src/libc/string.rs

```diff
@@ -142,23 +142,33 @@ fn strstr(env: &mut Environment, string: ConstPtr<u8>, substring: ConstPtr<u8>)
 
     let mut offset = 0;
     loop {
+        // Try to find the first matching character.
+
         let char_string = env.mem.read(string + offset);
 
         if char_string == 0 {
+            // We reached the end of the string, and no match was found.
             return Ptr::null(); // NULL
         }
 
         if char_string == char_substring {
-            let char_substring_plusone = env.mem.read(substring + 1);
-            let char_string_plusone = env.mem.read(string + offset + 1);
+            // We found a match - let's compare the rest of the substring to see if it matches too.
+
+            let mut inner_offset = 1;
 
             loop {
+                let char_substring_inner = env.mem.read(substring + inner_offset);
+                let char_string_inner = env.mem.read(string + offset + inner_offset);
+
                 if char_substring == 0 {
```

Comment:
Shouldn't this also check `char_substring_inner`?

I think this implementation is overcomplicated by trying to match the first character separately from the rest of the characters. It can probably be simplified to something like:

```rust
let mut offset = 0;
loop {
    let mut inner_offset = 0;
    loop {
        let char_string = env.mem.read(string + offset + inner_offset);
        let char_substring = env.mem.read(substring + inner_offset);
        if char_substring == b'\0' {
            return string + offset;
        } else if char_string == b'\0' {
            return Ptr::null();
        } else if char_string != char_substring {
            break;
        } else {
            offset += 1;
        }
    }
    offset += 1;
}
```

Note that I use `b'\0'` instead of `0` to emphasise it's a character, but that's just personal preference…

Hmm, I may as well commit that as my own code, I guess I'll do that once I've tested it a bit.

Reply:
That makes sense. I was trying to do an implementation similar to the one in the link I reference (https://opensource.apple.com/source/tcl/tcl-10/tcl/compat/strstr.c.auto.html), but this seems to be a simpler one.

Reply:
If it's derived from the algorithm in public source code, we shouldn't use it anyway due to the copyright risk.

Reply:
Did this now: 741c509061ee5ce3291cd18d0b81ddcfa237612b.